rust/pact_consumer API overhaul

emk opened the pull request:
Ready to merge (I hope)! ~~**Not yet ready to merge!** Please do not merge this branch yet; it's intended to provide an opportunity for ongoing feedback.~~

This PR contains my various API experiments proposed in https://github.com/pact-foundation/pact-reference/issues/17, plus a few other useful Rust tricks. I'm making an effort to keep each commit clear and independent, with a commit message describing the motivation by the change.

- [x] Create a top-level cargo workspace allowing us to run `cargo test --all` across all crates.
- [x] Make sure all the tests work.
- [x] Make a compilable and runnable version of the `pact_consumer` doctest, so that we have a metric by which to evaluate the API simplicity (and a fully working example for people using the library).
- [x] Remove the need to import `OptionalBody` directly.
- [x] Replace the `s!` macro with `Into<String>`.
- [x] Remove the need to pass hash tables to `headers` and `query`.
- [x] Figure out if we can avoid the need to wrap things in `OptionalBody`.
- [x] Think about the internal state machine, and whether or not it would be worth exposing this as part of the API types. (Diesel does something similar.) This interacts with the "builder" pattern in some important ways, though. (Don't worry, I'll explain this later!)
- [x] Provide an `Into<PactPattern>` interface (or something with a similar name), which allows the mixing of `String`, `&str`, `Regex`, `like(...)`, `serde_json::Value`, etc., when building interactions.
- [x] Experiment with some new ideas for the running & validation APIs.
- [x] Port unit tests elsewhere in the tree to the new API.

Outstanding API questions:

- [x] Do we want separate `StringPattern` and `JsonPattern` types? This would make it much easier to add `XmlPattern` someday. But it may add more complexity than we want. Experimentation required.
- [x] Q: Can we avoid needing to write `SomethingLike::new(json_pattern!("Archimedes"))` by using `Into` _without_ making it impossible to add `StringPattern` later? A: No.
- [x] What should we do about multiple values for a `query_param`? Getting a clean API for this with no hacks probably requires a bit of thought.
- [x] Add a `pact_term!` macro.
- [x] Add explicit support for `something_like` and `array_like` to the `json_pattern!` syntax.
- [x] Add a `url_for("/users/12")` method that calls `join` and `expect` with a relative path.
- [x] Add helpers like `content_type`, `json`, `not_found`, `created`, etc.
- [x] Q: Should we shorten the name of `json_pattern!` to something like `json_pat!`? A: Probably not.
- [x] Q: Should we rename `array_like!` (matches Ruby) to `each_like!` (matches JavaScript)? What about `something_like!` to `like!`?
- [x] Q: Do regex expressions all need `^` and `$`, or is that supplied by Ruby? A: Required on the Rust side.
- [x] Can we add `min` to the `array_like!` macro without too much messing around?
- [x] Add `charset=utf-8` (with case insensitive regex) to default JSON header.
- [x] Add helper to strip null object fields from `serde_json::Value`.
- [x] Figure out what's going on in the Ruby code, especially re: `min` and `match`.
- [x] ~~Add output directory configuration.~~ Will move to another PR.
- [x] Spent 10 Oct 2017 writing code against this API, fix any remaining issues, and submit this PR by EOD.

Other cleanup:

- [x] Add a helper method for `.get_or_insert_with(Default::default)`, and clean up everywhere.
- [x] Add an example testing with a domain object serialized to JSON.
- [x] Update top-level crate documentation to contain a brief explanation of what Pact is and why it's cool, for people who discover this crate at http://docs.rs/pact_consumer/

emk commented:
AppVeyor is breaking on this patch, mostly because it seems to be using an older Rust compiler that doesn't support the latest version of `itertools`:

```
   Compiling itertools v0.4.19
...
error: struct field shorthands are unstable (see issue #37340)
   --> C:\Users\appveyor\.cargo\registry\src\github.com-1ecc6299db9ec823\unicode-bidi-0.3.4\src\explicit.rs:180:32
    |
180 |         self.vec.push(Status { level, status });
    |                                ^^^^^
error: struct field shorthands are unstable (see issue #37340)
```

I'm going to leave this alone for now, and just make sure that this PR builds on Linux while I'm working on it.

emk commented:
### Experiments so far

OK, I've pushed my first batch of changes! Here's a summary of some highlights.

#### 1. Improved development workflow for `pact-reference` crates.

You can now run all tests for all crates by `cd`ind to `pact-reference/rust` and typing:

```sh
cargo test --all
```

Or you can run:

```sh
cargo install cargo-watch
cargo watch -x 'test -all'
```

...for automatic recompilation of all code, and re-running of all test suites on each save. The `--all` flag uses the relatively recent Cargo `[workspace]` feature, and it allows sharing compilation artifacts between a related family of crates, for much faster testing performance.

#### 2. Experimental ergonomic tweaks to `pact_consumer`
 
The current `pact_consumer` API looks like this:

```rust
    let pact_runner = ConsumerPactBuilder::consumer(s!("Consumer"))
        .has_pact_with(s!("Alice Service"))
        .given("there is some good mallory".to_string())
        .upon_receiving("a retrieve Mallory request".to_string())
            .path(s!("/mallory"))
        .will_respond_with()
            .status(200)
            .headers(hashmap!{ s!("Content-Type") => s!("text/html") })
            .body(OptionalBody::Present(s!("That is some good Mallory.")))
        .build();
```

With a bit of Rust tweaking, we can make it look like this:

```rust
    let pact_runner = ConsumerPactBuilder::consumer("Consumer")
        .has_pact_with("Alice Service")
        .given("there is some good mallory")
        .upon_receiving("a retrieve Mallory request")
            .path("/mallory")
        .will_respond_with()
            .status(200)
            .header("Content-Type", "text/html")
            .body_present("That is some good Mallory.")
        .build();
```

Note that this is still (mostly!) compatible with the existing API. You can pass `&str` or `String` or even `std::borrow::Cow<str>`, and it should do the right thing. This is thanks to the magic of `Into<String>`, which allows any type that the compiler knows how to convert to a `String`. (Not every Rust API should use `Into<String>`, because it adds complexity, but it works especially well for "builder" APIs like this one.)

### Where to next?

I feel like the block:

```
        .will_respond_with()
            .status(200)
            .header("Content-Type", "text/html")
            .body_present("That is some good Mallory.")
```

...could be made more ergonomic with further tweaks:

1. Should we set `status(200)` as a default? This is what people will specify the most often. But I don't understand the underlying match algorithm, so maybe this is a bad idea?
2. We could combine `header` with `Content-Type` and `body` into helper methods like `json_body` for common cases.
3. We might want to rename `body(&mut self, body: OptionalBody)` to `optional_body`, and re-use the name `body` for `body_present`, because that's going to be the most common case by far. That would allow a more consistent naming scheme for things like `body`, `json_body`, etc.

Any thoughts on the best way to tackle these issues? I don't understand matching at all yet, so I'm probably getting some critical stuff wrong! 

After this, the next step is to add support for regular expressions, `serde_json::Value` literals, and helper functions such as `like`. Before I tackle this, I need to skim through the matching algorithm and look at the other language implementations, to make sure I understand everything.

Thank you for any ideas and feedback you provide (and for such a handy library)!

emk commented:
Progress!

I've implemented `JsonPattern` and `Term` types (borrowing both ideas and some macro code from `serde_json`, with all licensing info properly intact), and I can now parse things like:

```rust
json_pattern!({
    "message": "Hello, world!",
    "location": { "x": 1, "y": 2 },
    "tags": ["interesting"]
});

json_pattern!({
    // You can use Rust expressions, as long as they support
    // `Into<PatternJson>`.
    "message": format!("Hello, {}!", "world"),

    // You can also nest the `json!` macro to embed types which
    // support `Serialize`.
    "location": json!(Point { x: 1.0, y: 2.0 }),

    // You can use the various `Term` functions to match values by
    // by type, length or regular expression.
    "tags": Term::like(json_pattern!(["tag"])),
});
```

There's obviously a whole bunch of Rust magic under the hood to make this work, including plenty of `Into<JsonPattern>` implementations for common Rust types. This means that strings, numbers, booleans, etc., are all automatically promoted to `JsonPattern` within the body of the macro. So we have the basic skeleton of a Rust DSL.

Still remaining:

- Actually generate JSON path expressions and the JSON specifications for matchers.
- Hook this into the `body` API so that we can conveniently specify JSON bodies.
- Figure out how to use `Term` expressions for non-`body` data.

Anyway, I'm really interested in feedback! Does this seem to be headed in a useful direction? Have I misunderstood any key Pact ideas?

mefellows commented:
I know nothing of Rust, I'm just happy to see other clever people getting on board and wanting to contribute - so i'll continue to watch from afar until such time that my desk is clear and I can learn a new language :)

emk commented:
@mefellows Thank you for the encouragement!

@uglyog @mboudreau I'm reaching a point where I'm pretty happy with the `json_pattern!` macro, and I'm ready to integrate it into my copy of the  `ConsumerPactBuilder`. I've also been writing quite a few sample tests, including some for our production code.

And the more I play around with Rust Pact code, the more I'm leaning towards the idea of changing the builder API a bit. Let me explain. :-)

### The published API

Here's what the current API looks like:

```rust
    let pact_runner = ConsumerPactBuilder::consumer(s!("Consumer"))
        .has_pact_with(s!("Alice Service"))
        .given("there is some good mallory".to_string())
        .upon_receiving("a retrieve Mallory request".to_string())
            .path(s!("/mallory"))
        .will_respond_with()
            .status(200)
            .headers(hashmap!{ s!("Content-Type") => s!("text/html") })
            .body(OptionalBody::Present(s!("That is some good Mallory.")))
        .build();
```

Some interesting details to note here:

1. We're building four separate kinds of objects here: a `Pact`, `Interaction`s, `Request`s and `Response`s. These are all handled by a single `ConsumerPactBuilder` class, using an internal `BuilderState` state machine.
2. Certain calls (like `upon_receiving` or `path`) will automatically push the current interaction (if appropriate) and start a new one. Other calls (like `will_response_with` and `response`) switch the state machine so that `body` applies to the `Response` instead of the `Request`.
3. Certain parts of this syntax may contain surprises. For example, `given` only applies to the next interaction, and a second `upon_receiving` will reset the `given` to empty.

### My "work-in-progress" API

This is basically the same, except that it tries to minimize the use of macros like `s!` and `hashmap!`, and it has partial support for the `json_pattern!` macro:

```rust
    let pact_runner = ConsumerPactBuilder::consumer("Consumer")
        .has_pact_with("Alice Service")
        .given("there is some good mallory")
        .upon_receiving("a retrieve Mallory request")
            .path("/mallory")
            .query_param("quality", "good")
        .will_respond_with()
            .status(200)
            .header("Content-Type", "text/html")
            .body("That is some good Mallory.")
        .build();
```

This is mostly source compatible with the old API. A few interesting details have changed:

1. I've added `header` and `query_param` functions that allow adding a single header or parameter at time.
2. It's possible to write something like `.path(Term::new(greeting_regex, "/greeting/hello"))` and get a working `{ "match": "regex", "regex": ... }` clause generated. If I finished implementing this, I would make the `json_pattern!` macro available everywhere that it's allowed.

OK, now onto some of the issues I've been wrestling with...

### Problem #1: `rustfmt` is the "standard Rust style", and it enforces how builders will be indented

Rust provides a [`rustfmt`](https://github.com/rust-lang-nursery/rustfmt) tool which converts Rust code to using a "standard Rust style". This standard style is determined via a [public RFC process](https://github.com/rust-lang-nursery/fmt-rfcs).

Unfortunately, `rustfmt` breaks the nice two-level indentation in original build. This:

```rust
        .upon_receiving("a retrieve Mallory request")
            .path("/mallory")
            .query_param("quality", "good")
        .will_respond_with()
            .status(200)
            .header("Content-Type", "text/html")
            .body("That is some good Mallory.")
        .build();
```

...becomes this:

```rust
        .upon_receiving("a retrieve Mallory request")
        .path("/mallory")
        .query_param("quality", "good")
        .will_respond_with()
        .status(200)
        .header("Content-Type", "text/html")
        .body("That is some good Mallory.")
        .build();
```

This is a lot less readable unless we insert blank lines. Ideally, it would be nice to avoid this problem by choosing a builder syntax that works well after being run through `rustfmt`.

### Problem #2: The `BuilderState` state machine makes the implementation fairly complex, and my new features make it worse

The `BuilderState` type keeps track of whether we're building a request, a response, or nothing:

```rust
enum BuilderState {
    None,
    BuildingRequest,
    BuildingResponse
}
```

This is used to implement the state machine behind `path` and `will_respond_with`, as well as the logic for when to start building a new `Interaction`. Behind the scenes, this involves some moderately complex code:

```rust
    /// Describe the state the provider needs to be in for the pact test to be verified. (Optional)
    pub fn given(&mut self, provider_state: String) -> &mut Self {
        match self.state {
            BuilderState::None => (),
            _ => self.pact.interactions.push(self.interaction.clone())
        }
        self.interaction = Interaction {
            provider_state: Some(provider_state.clone()),
            .. Interaction::default()
        };
        self.state = BuilderState::BuildingRequest;
        self
    }
```

...or:

```rust
    /// Headers to be included in the request
    pub fn headers(&mut self, headers: HashMap<String, String>) -> &mut Self {
        match self.state {
            BuilderState::BuildingRequest => self.interaction.request.headers = Some(headers.clone()),
            BuilderState::BuildingResponse => self.interaction.response.headers = Some(headers.clone()),
            BuilderState::None => {
                self.interaction.request.headers = Some(headers.clone());
                self.state = BuilderState::BuildingRequest;
            }
        };
        self
    }
```

If this were the only complexity, it wouldn't be a deal-breaker. But once I started add support for `matching_rules`, this state machine started interacting very badly with the borrow checker, because I need to borrow _both_ `&mut request.headers` and `&mut request.matching_rules`, which leads to terrible Rust madness like this:

```rust
    /// Internal function which returns a mutable version of our query
    /// and a mutable refence to our current collection of matching rules.
    fn query_mut(&mut self) -> (
        &mut HashMap<String, Vec<String>>,
        &mut Matchers,
     ) {
        self.push_interaction();
        let req = &mut self.interaction.request;
        (
            // This is a fairly tricky bit of Rust, because we need to borrow
            // two fields of `req` mutably. Rust will let us do this, but it
            // needs to see both borrows in the same function, and not hide
            // either of them elsewhere.
            req.query.get_or_insert_with(Default::default),
            req.matching_rules.get_or_insert_with(Default::default),
        )
    }
```

...and something _vastly_ worse for `headers_mut`, because it needs to deal with every possible `BuilderState`. I spent like two hours this morning trying to get `headers_mut` past the borrow checker, and it's an ugly nightmare.

The first rule of Rust Happiness is "If you're getting into huge fights with the borrow checker, then back off and simplify the mutable state in your design." So I decided to try that, just to see what would happen.

### A possible alternative API

After much experimentation, I managed to create an API that looks like this:

```rust
    let pact = PactBuilder::new("Consumer", "Provider")
        .interaction("GET /greeting/hello", |i| {
            i.given("a greeting named hello");
            i.request.method("GET").path("/greeting/hello");
            i.response
                .status(200)
                .header("Content-Type", "application/json")
                .json_body(json_pattern!({
                    "message": "Hello!",
                }));
        })
        .build();
```

Here are some interesting details:

1. There is no more `BuilderState`! Instead, we build our `Pact` with `PactBuilder`, our `Interaction`s with an `InteractionBuilder`, our `Request`s with a `RequestBuilder` and our `Response`s with a `ResponseBuilder`. This removes a ton of implementation complexity.
2. The `interaction` constructs a new `InteractionBuilder` for you, and allows to access all the builders for that interaction. When done, your `Interaction` is added to the `Pact`. This also helps `rustfmt` get the formatting right! (Everything above has been formatted with `rustfmt`.)
3. Mandatory arguments are all passed as regular function parameters (to `new`, etc.). Optional parameters are represented as builder methods.
4. We now build a regular `Pact`, and not a `ConsumerPactRunner`, which makes the API much easier to test, and which allows reuse in more contexts.

### What the new implementation looks like

The implementation of this code is slightly longer, but it's actually a lot simpler, and it can be extended without the kind of more advanced Rust knowledge I needed in `query_mut`. Here's a sample. (Note the use of `Into<String>` to allow passing `String`, `&str`, `Cow<str>`, etc.)

```rust
struct PactBuilder {
    pact: Pact,
}

impl PactBuilder {
    fn new<C, P>(consumer: C, provider: P) -> Self
    where
        C: Into<String>,
        P: Into<String>,
    {
        let mut pact = Pact::default();
        pact.consumer = Consumer { name: consumer.into() };
        pact.provider = Provider { name: provider.into() };
        PactBuilder { pact: pact }
    }

    fn interaction<D, F>(&mut self, description: D, build_fn: F) -> &mut Self
    where
        D: Into<String>,
        F: FnOnce(&mut InteractionBuilder),
    {
        let mut interaction = InteractionBuilder::new(description.into());
        build_fn(&mut interaction);
        self.push_interaction(interaction.build())
    }

    fn push_interaction(&mut self, interaction: Interaction) -> &mut Self {
        self.pact.interactions.push(interaction);
        self
    }

    fn build(&self) -> Pact {
        self.pact.clone()
    }
}
```

As you can see, this is pretty vanilla Rust code, with nothing too exciting going on (except for the `FnOnce` type).

### Possible drawbacks

This new API would require breaking source compatibility (and releasing an 0.3.x release, if we go by Rust community conventions). In certain circumstances, it might use an extra line or two of vertical space, and it no longer allows things like omitting `upon_receiving` and relying on the state machine to do the right thing.

Anyway, what do you think? I'd really love some feedback on the design, especially in places where I've misunderstood Pact semantics. Would you be interested in having an API like for `pact_consumer`? If so, I can have a working PR for you pretty quickly.

Anyway, thank you for all your encouragement! (And I'm looking forward to moving one of Faraday's critical production tools over to Pact, so that I do a major refactoring!)

emk commented:
OK, I've pushed an implementation of the API described in "A possible alternative API" to the https://github.com/faradayio/pact-reference/tree/rust_api_ideas_wip branch. I would still need to fix a couple of tests in other crates and update the `ConsumerPactRunner` API to work with the new `PactBuilder`, but the hard part is all done.

Here's an example of the new API from the documentation:

```rust
let pact = PactBuilder::new("Greeting Client", "Greeting Server")
    .interaction("asks for a greeting", |i| {
        i.request.path("/greeting/hello");
        i.response
            .header("Content-Type", "application/json")
            .json_body(json_pattern!({ "message": "hello" }));
    })
    .build();
```

emk commented:
### Another interesting use case

The existing `pact_consumer` API has a `run` parameter which takes a block argument.

```rust
let pact_runner = ConsumerPactBuilder::consumer(s!("Consumer"))
    // ...
    .build();

let result = pact_runner.run(&|url| {
    // You would use your actual client code here
    let client = Client { url: url.clone(), .. Client::default() };
    // we get our client code to execute the request
    let result = client.fetch("/mallory");
    expect!(result).to(be_ok().value("That is some good Mallory."));
    Ok(())
});
// This means it is all good
expect!(result).to(be_equal_to(VerificationResult::PactVerified));
```

Two things worth noting:

1. If your client needs to talk to two different servers, you need to nest two `run` blocks. This can get messy.
2. You need to remember to put `expect!(result).to(be_equal_to(VerificationResult::PactVerified));` at the end of each test, or else the tests don't check anything.

### An alternative API

I experimented with several alternatives. One interesting approach is as follows:

```rust
#[test]
fn mock_server_passing_validation() {
    // Define the Pact for the test, specify the names of the consuming
    // application and the provider application.
    let alice_service = PactBuilder::new("Consumer", "Alice Service")
        // Start a new interaction. We can add as many interactions as we want.
        .interaction("a retrieve Mallory request", |i| {
            // Defines a provider state. It is optional.
            i.given("there is some good mallory");
            // Define the request, a GET (default) request to '/mallory'.
            i.request.path("/mallory");
            // Define the response we want returned.
            i.response
                .status(200)
                .header("Content-Type", "text/plain")
                .body("That is some good Mallory.");
        })
        .start_mock_server();

    // You would use your actual client code here.
    let mallory_url = alice_service.url().join("/mallory").unwrap();
    let mut response = reqwest::get(mallory_url).expect("could not fetch URL");
    let mut body = String::new();
    response.read_to_string(&mut body).expect("could not read response body");
    assert_eq!(body, "That is some good Mallory.");

    // When your test has finished running, all verifications will be performed
    // automatically, and an error will be thrown if any have failed.
}
```

This approach represents each mock server with a value like `alice_service`. When this value goes out of scope, the server is automatically stopped and all validations are run, thanks to an `impl Drop`. (There's some code in there to avoid panicking with a validation error while Rust is already panicking about a client code error.)

### It's working, and all tests pass!

I've implemented all of my experimental API changes, written lots of new tests, and made sure that all the existing tests pass for all the other Pact modules in the tree. In particular, the `rust/pact_verifier/src/pact_broker.rb` file contains quite a few complicated examples, all of which I've ported to the experimental API.

As always, I'm very interested in feedback! The first post in this thread has a checklist of minor API issues that I'm still thinking about, and I'd be happy to make other improvements, too. Once again, thank you for the encouragement.

(And now I'm going to go off and write a bunch of Pact code for one of our Rust production systems, and see how this new API works in practice.)

uglyog commented:
This is the most well documented PR I've ever reviewed! :-D

I've worked my way through your comments, and I'm going to review the actual code changes next, but from what I've seen so far, I'm happy to merge this in when you are ready. What you have done is awesome, as it makes it much more idiomatic Rust code. I had just converted the original from the JVM version.

I'll update the Apveyor build to hopefully get that to pass.

uglyog commented:
I have updated the Appveyor build to use rustup to install the Rust toolchain, so if you merge master in it should start running the build and hopefully pass.

mefellows commented:
> This is the most well documented PR I've ever reviewed! :-D

Indeed! So well thought through and crafted. I'm going to look at transferring some of this thinking the the JS and Go implementations too (Go does some similar things already). 

Really well done!

emk commented on rust/pact_consumer/src/mock_server.rs:
```diff
@@ -0,0 +1,140 @@
+        if mismatches.is_empty() {
+            // Success! Write out the generated pact file.
+            lookup_mock_server_by_port(self.port, &|ms| {
+                ms.write_pact(&Some("target/pacts".to_owned()))
```
OK, cool! I just copied this code straight over from the original, but I'm happy to add a configuration scheme. How would you like this to work? An environment variable?

uglyog replied:
An environment variable would be acceptable.

emk commented:
@uglyog @mefellows Thank you for the kind words!

I spent today writing test code using my branch of `pact_consumer`, and I polished off some rough edges. I also went through the TODO list (in the top comment), and made a whole bunch of minor improvements.

Some highlights:

- In addition to `JsonPattern`, I also added a `StringPattern` type. This is used for things like `path` and `headers`, to restrict the values to ones which make sense for strings. This simplifies the code, and it enforces correct API usage at compile time. It also adds most of the hooks that would be needed to implement an `XmlPattern` type in the future, if anybody is feeling really ambitious. :-)
- I added new macros for `term!`, `array_like!` and `something_like!`, which really cleans up the syntax in day-to-day use. In particular, it should no longer be necessary to nest `json_pattern!` calls everywhere.
- I added lots of little helper methods like `.post()` and `.content_type(...)` to the builders, making it easier to handle common cases.

A complex example now looks like:

```rust
use pact_consumer::prelude::*;

PactBuilder::new("quotes client", "quotes service")
    .interaction("add a new quote to the database", |i| {
        i.request
            .post()
            .path("/quotes")
            .json()
            .json_body(json_pattern!({
                 // Allow the client to send any string as a quote.
                 // When testing the server, use "Eureka!".
                 "quote": something_like!("Eureka!"),
                 // Allow the client to send any string as an author.
                 // When testing the server, use "Archimedes".
                 "by": something_like!("Archimedes"),
                 // Allow the client to send an array of strings.
                 // When testing the server, send a single-item array
                 // containing the string "greek".
                 "tags": array_like!("greek"),
             }));

        i.response
            .created()
            // Return a location of "/quotes/12" to the client. When
            // testing the server, allow it to return any numeric ID.
            .header("Location", term!("^/quotes/[0-9]+$", "/quotes/12"));
    });
```

With the new macros, it's much easier to easier to build complex, nested patterns:

```rust
json_pattern!({
  // Expect an array of strings.
  "tags": array_like!("tag"),

  // Expect an array of objects, each of which has a name key containing
  // a string (but match the actual names by type).
  "people": array_like!({
    "name": "J. Smith",
  }),
});
```

I think this is getting pretty close to it's final form. I'm going to spend tomorrow writing lots of unit tests in Rust, and making sure there are no other places that need polishing. Once that's done, I'll give you the go ahead tomorrow and we can merge this then.

I may still be misunderstanding critical ideas about how Pact works, so I really appreciate any feedback you can provide!

uglyog commented:
You have definitely grasped the critical ideas about how Pact works, and all your changes make perfect sense. If something stands out, I'll let you know, but at this point I'm happy to merge it and then fix anything that comes up.

emk commented:
Give me one more day to break APIs if necessary, and then we can go ahead and merge it!

emk commented:
OK, looks like a I found a couple of issues when I tested against `pact-js`:

1. `pact_matching` wasn't handling JSON paths containing underscores, such as `$.query.user_id`. I've added a fix for this.
2. I'm not generating working matches for `array_like`.

```
[Error: WARN: Ignoring unsupported matching rules {"match"=>"type", "min"=>"1"} for path $['body']
```

emk commented:
Good news! I've managed to run pacts generated by the new Rust `pact_consumer` application against `pact-js` and an in-house Node app. It's pretty cool, but it turns out that I need to tweak a couple of helper functions to make things painless for users.

I do have a couple of API questions:

- Should I rename `array_like!` (matches Ruby) to `each_like!` (matches JavaScript), and `something_like!` to `like!`? I think the JavaScript versions are clearer, but I also want to be consistent with cross-language Pact APIs whenever possible.
- When I write `term!("^/users/[0-9]+$", "/users/12")`, is it safe to leave out the `^` and `$`? Does the Pact specification assume that all `{ "type": "regex" }` matchers should be treated as fully "anchored" expressions?

Thank you very much for your feedback!

emk commented:
OK, I went ahead and renamed `array_like!` (Ruby style) to `each_like!` (JS style), etc., and the APIs look nicer, I think. I also verified that `^` and `$` are mandatory in regular expressions, at least if we're using the Rust matching engine.

I've gone through and made almost all the other tweaks and cleanups I wanted, leaving only two issues:

1. I need to review the implementation of `each_like!` to make sure that I generate the same `matching_rules` as Ruby.
2. I still need to add support for configuring the output directory. This is actually a new feature, but I'm happy to do it! But it probably doesn't need to block the merge, strictly speaking, since it's not a regression.

Let me take care of (1), run a final test with our in-house code, and then this PR will be mergeable!

emk commented:
I have two questions before we merge. :-)

I made one last check to make sure I got all the matching rules correct. I'm checking against Ruby, and I'm pretty sure I generate the same output.

### A possible bug in the Ruby code

If you look at the [Ruby version of the matching rules](https://github.com/pact-foundation/pact-support/blob/master/lib/pact/matching_rules/extract.rb), there's a `match_type` parameter that gets passed down recursively:

```ruby
      def recurse object, path, match_type
        case object
        when Hash then recurse_hash(object, path, match_type)
        when Array then recurse_array(object, path, match_type)
        when Pact::SomethingLike then handle_something_like(object, path, match_type)
        when Pact::ArrayLike then handle_array_like(object, path, match_type)
        when Pact::Term then record_regex_rule object, path
        when Pact::QueryString then recurse(object.query, path, match_type)
        when Pact::QueryHash then recurse_hash(object.query, path, match_type)
        end
    end
```

This gets set to `:array_like` here:

```ruby
      def handle_array_like array_like, path, match_type
        record_rule "#{path}", 'min' => array_like.min
        record_match_type_rule "#{path}[*].*", 'type'
        recurse array_like.contents, "#{path}[*]", :array_like
    end
```

...and it gets checked against `:array_like` here:

```ruby
      def record_match_type_rule path, match_type
        unless match_type == :array_like || match_type.nil?
          rules[path] ||= {}
          rules[path]['match'] = match_type
        end
      end
```

But here's the problem: I can't find any way to get `:array_like` from `handle_array_like` to `record_match_type_rule`. As far as I can tell, the `match_type` parameter to `recurse` is ignored on all code paths.

Could somebody please look at the Ruby code for me and let me know what's going on?

### Does `"min"` require `"match": "type"`?

Sometimes, the Pact specs mention `[{"min": 1, "match": "type"}]`, and sometimes they mention `[{"min": 1}]`. What's the difference between these two? Here's an example from the V3 spec:

```json
{
  "body": {
    "$.animals": {
        "matchers": [{"min": 1, "match": "type"}]
    },
    "$.animals[*].*": {
        "matchers": [{"match": "type"}]
    },
    "$.animals[*].children": {
        "matchers": [{"min": 1}]
    },
    "$.animals[*].children[*].*": {
        "matchers": [{"match": "type"}]
    }
  }
}
```

These should be my last questions, with any luck. We're almost ready to merge!

emk commented:
I found a code difference between Rust's `pact_matching` and `pact-js`.

The Rust version generates matches which look like this, with a string for `min`:

```json
{ "match": "type", "min": "1" }
```
+            "min": "1"
           },
```

The JavaScript version (which I think is just calling Ruby), however, insists on:

```json
{ "match": "type", "min": 1 }
```

...with an actual number.

I assume the numeric version is correct. What about the string version? Is that ever sent by any of the other languages?

I'm happy to include a fix for `pact_matching`, but I'd appreciate any help you can provide in figuring out how it should behave. :-)

bethesque commented:
I reckon an int rather than a string.

bethesque commented:
Two things: this "matchers" element, that's different from the ruby and the jvm (unless the v3/v4 specs have changed significantly, @uglyog??)

Currently, ruby generates this for an `each like`

```
"matchingRules": {
  "$.body.factories": {
    "min": 1
  },
  "$.body.factories[*].*": {
    "match": "type"
  }
}
```

I've been convinced that the match should just go on the parent like:
```json
"matchingRules": {
  "$.body.factories": {
    "min": 1,
    "match": "type"
  }
}
```

I can't remember, is that what jvm does already Ron?

uglyog commented:
The matchers above is the new V3 format.

Just a note that the Rust matching is based on the JVM one, and we know there are some differences between the JVM implementation and the Ruby one, so these should be expected.

I also want to state that the Ruby implementation was the first one we built, so it may not be the best one to use as a baseline. We should take each difference and work out what is best on a case by case basis.

For matching issues, let us raise new issues rather than try fix them as part of this PR, as it may cause this PR to drag on. Same goes for configuring the output directory, we could do that in a separate issue.

### Does "min" require "match": "type"?

The answer here is no, but it is generally used with type matching, that is why they are normally together.  I just checked the JVM version, and it will handle just the `min` by itself by assuming the `"match": "type"` part. It will also handle both `"min": "1"` and `"min": 1`.

What the JVM generates depends on what the example is set as, it does not automatically add the last wildcard (`.*` part). The  `"match": "type"` cascades, so it is added to the parent and inherited by the children.

For this example

```groovy
ids minLike(1) {
    id integer(100)
}
```

It will generate the following two rules:

```json
"matchingRules": {
  "$.body.ids": {
    "min": 1,
    "match": "type"
  },
  "$.body.ids[*].id": {
    "match": "integer"
  }
}
```

emk commented:
We're agreed on:

- We should put the output directory changes into a separate PR.
- Rust should accept a missing `matches` parameter with `max` and `min`, but it should always generate `"matches": "type"`.
- We can fine-tune matching in another PR.

This PR has one major interoperability bug that prevents testing the generated pacts against providers written in other languages:

- The Rust `pact_matching` library generates `{ "match": "type", "min": "1" }` instead of `{ "match": "type", "min": 1 }`, which other languages rightfully refuse to load. I've filed a separate issue at #20 to figure out how to fix this.

My recommend: Merge this PR, but _don't_ release until we fix #20, which is needed to interop with other languages. Anybody else have any final thoughts?

Thank you to everybody for all your help and feedback, and for being willing to accept such a far-reaching patch!

emk commented:
I've filed new issues for the following:

- #21: Configurable output directory.
- #22: Should `each_like!` have magic matching rules for the keys of child objects like Ruby does, or should it use another macro for that?

Please feel free to add comments there!

uglyog commented:
Ok, PR merged! Thanks @emk for your contribution. I'll get on to #20, it should be a small change to fix.
